### PR TITLE
torcontrol: Launch a private Tor instance when not already running

### PR DESCRIPTION
**Pull request by luke-jr**

### CODE REVIEW

**Comment by luke-jr** [comment]
NOTE: This uses boost::process, but currently doesn't have anything to check that it's actually available.

**Comment by practicalswift** [comment]
Concept ACK assuming it is done without using `boost::process`. Not worth the extra dependency.

Needs a test: Could use a mocked `tor` command.

**Comment by Sjors** [comment]
@luke-jr #15382 in its current incarnation uses `boost::process` as well, and includes some simple detection code (it just checks if the version is >=1.64).

@practicalswift I agree we shouldn't be launching arbitrary binaries named "tor" without the user opting into that. I'm not sure how this PR changes [existing behavior](https://github.com/bitcoin/bitcoin/blob/master/doc/tor.md). For modern versions of Tor we automatically create a hidden service if the user is running Tor: "if Tor is running (and proper authentication has been configured), Bitcoin Core automatically creates a hidden service to listen on", but that's quite different from _starting_ Tor.

**Comment by practicalswift** [comment]
@Sjors Yes, that is very different: if Tor is already running and proper authentication has been configured then the user has already opted in Tor and can be assumed to know the consequences of doing so.

File: src/torcontrol.cpp [hunk 1]
```diff
@@ -681,22 +705,84 @@ void TorController::protocolinfo_cb(TorControlConnection& _conn, const TorContro
...
+    std::remove(controlport_env_filepath.c_str());
+
+    if (m_process) {
+        m_process->terminate();
+        delete m_process;
+    }
+
+    boost::process::opstream in;
+    try {
+        m_process = new boost::process::child(m_execute + " -f -", boost::process::std_in < in);
```
Comment (practicalswift) [inline src/torcontrol.cpp]: `unique_ptr`?

**Comment by luke-jr** [comment]
@practicalswift 

> This should be opt-in and not enabled by default, right?

> Starting tor without the end-user asking for it explicitly is highly surprising. Also it could cause potentially unwanted outbound TCP connections to be made.

We already make outbound TCP connections by default. Note that this does not operate Tor as an exit node, nor does it even use it for outbound connections - it is exclusively for an inbound hidden service only.

The end goal, is to display the hidden service address in a QR code for easy pairing with mobile wallets.

> Please avoid introducing the boost::process dependency.

Not practical. We're going to need it either way from the sound of it.

**Comment by luke-jr** [comment]
configure now checks for boost::process availability, and `-onion` is not configured for private Tor instances (which don't offer the SOCKS service).

**Comment by practicalswift** [comment]
@luke-jr Could you describe what benefits auto-starting Tor brings to the end-user?

What are the arguments against having the end-user opt in to this behaviour?

> We already make outbound TCP connections by default.

Yes, but we make outbound TCP connections on port 8333 to the Bitcoin network, which is something the end-user can reasonably assume will happen when starting `bitcoind`.

Auto-starting Tor and thus establishing outbound TCP connections to the Tor network is another thing entirely.

**Comment by practicalswift** [comment]
Could add `FascistFirewall 1` to the config to make sure only port 80 and 443 are used?

**Comment by luke-jr** [comment]
@practicalswift I don't see the need - we already connect out on 8333...

**Comment by practicalswift** [comment]
@luke-jr Sorry for being unclear. I was referring to the Tor network traffic :-)

**Comment by luke-jr** [comment]
@practicalswift What does it matter?

**Comment by Sjors** [comment]
> Note that this does not operate Tor as an exit node, nor does it even use it for outbound connections - it is exclusively for an inbound hidden service only.

That's still detectable for network observers afaik. Let's just add a `tor=` setting.

It also seems unsafe to call a random executable in the users path called `tor`, so we could ask for a full path, but that might be too annoying UX. The `tor=` setting could help in that regard too.

**Comment by practicalswift** [comment]
@luke-jr I don't hold any strong position but my reasoning was the following: if we're adding a default configuration then we might as well doing it using port settings which maximise the likelihood of it working out of the box.

**Comment by luke-jr** [comment]
Rebased

**Comment by luke-jr** [comment]
Closing due to lack of agreement/reviews...

**Pull request closed without merging**